grudge/op.py

```diff
@@ -403,155 +995,133 @@ def norm(dcoll, vec, p, dd=None):
+        denote the element-wise sum of *vec*.
+    """
 
-def _cross_rank_trace_pairs_scalar_field(dcoll, vec, tag=None):
-    if isinstance(vec, Number):
-        return [TracePair(BTAG_PARTITION(remote_rank), interior=vec, exterior=vec)
-                for remote_rank in connected_ranks(dcoll)]
+    if len(args) == 1:
+        vec, = args
+        dd = dof_desc.DOFDesc("vol", dof_desc.DISCR_TAG_BASE)
+    elif len(args) == 2:
+        dd, vec = args
     else:
-        rbcomms = [_RankBoundaryCommunication(dcoll, remote_rank, vec, tag=tag)
-                for remote_rank in connected_ranks(dcoll)]
-        return [rbcomm.finish() for rbcomm in rbcomms]
-
-
-def cross_rank_trace_pairs(dcoll, ary, tag=None):
-    r"""Get a list of *ary* trace pairs for each partition boundary.
-
-    For each partition boundary, the field data values in *ary* are
-    communicated to/from the neighboring partition. Presumably, this
-    communication is MPI (but strictly speaking, may not be, and this
-    routine is agnostic to the underlying communication, see e.g.
-    _cross_rank_trace_pairs_scalar_field).
+        raise TypeError("invalid number of arguments")
 
-    For each face on each partition boundary, a :class:`TracePair` is
-    created with the locally, and remotely owned partition boundary face
-    data as the `internal`, and `external` components, respectively.
-    Each of the TracePair components are structured like *ary*.
+    dd = dof_desc.as_dofdesc(dd)
 
-    The input field data *ary* may be a single
-    :class:`~meshmode.dof_array.DOFArray`, or an object
-    array of ``DOFArray``\ s of arbitrary shape.
-    """
-    if isinstance(ary, np.ndarray):
-        oshape = ary.shape
-        comm_vec = ary.flatten()
-
-        n, = comm_vec.shape
-        result = {}
-        # FIXME: Batch this communication rather than
-        # doing it in sequence.
-        for ivec in range(n):
-            for rank_tpair in _cross_rank_trace_pairs_scalar_field(
-                    dcoll, comm_vec[ivec]):
-                assert isinstance(rank_tpair.dd.domain_tag, dof_desc.DTAG_BOUNDARY)
-                assert isinstance(rank_tpair.dd.domain_tag.tag, BTAG_PARTITION)
-                result[rank_tpair.dd.domain_tag.tag.part_nr, ivec] = rank_tpair
-
-        return [
-            TracePair(
-                dd=dof_desc.as_dofdesc(
-                    dof_desc.DTAG_BOUNDARY(BTAG_PARTITION(remote_rank))),
-                interior=make_obj_array([
-                    result[remote_rank, i].int for i in range(n)]).reshape(oshape),
-                exterior=make_obj_array([
-                    result[remote_rank, i].ext for i in range(n)]).reshape(oshape)
-                )
-            for remote_rank in connected_ranks(dcoll)]
-    else:
-        return _cross_rank_trace_pairs_scalar_field(dcoll, ary, tag=tag)
+    if isinstance(vec, np.ndarray):
+        return sum(elementwise_sum(dcoll, dd, vec_i) for vec_i in vec)
```

Comment:
@inducer Actually, looking at this now, I think this should be
```python
    if isinstance(vec, np.ndarray):
        return obj_array_vectorize(
            lambda vi: elementwise_sum(dcoll, dd, vi), vec
        )
```
What do you think?

Reply:
I'm pretty sure this is what it should be. So I'm going to add it in.